test/sslapitest.c

```diff
@@ -8045,7 +8046,193 @@ static int test_ssl_dup(void)
 
     return testresult;
 }
-#endif
+
+# ifndef OPENSSL_NO_DH
+/* Helper function for the test_set_tmp_dh() tests */
+static EVP_PKEY *get_tmp_dh_params(void)
+{
+    BIGNUM *p = NULL;
+    OSSL_PARAM_BLD *tmpl = NULL;
+    EVP_PKEY_CTX *pctx = NULL;
+    OSSL_PARAM *params = NULL;
+    EVP_PKEY *dhpkey = NULL;
+
+    p = BN_get_rfc3526_prime_2048(NULL);
+    if (!TEST_ptr(p))
+        goto end;
+
+    pctx = EVP_PKEY_CTX_new_from_name(libctx, "DH", NULL);
+    if (!TEST_ptr(pctx)
+            || !TEST_true(EVP_PKEY_key_fromdata_init(pctx)))
+        goto end;
+
+    tmpl = OSSL_PARAM_BLD_new();
+    if (!TEST_ptr(tmpl)
+            || !TEST_true(OSSL_PARAM_BLD_push_BN(tmpl,
+                                                    OSSL_PKEY_PARAM_FFC_P,
+                                                    p))
+            || !TEST_true(OSSL_PARAM_BLD_push_uint(tmpl,
+                                                    OSSL_PKEY_PARAM_FFC_G,
+                                                    2)))
+        goto end;
+
+    params = OSSL_PARAM_BLD_to_param(tmpl);
+    if (!TEST_ptr(params)
+            || !TEST_true(EVP_PKEY_fromdata(pctx, &dhpkey, params)))
+        goto end;
+
+ end:
+    BN_free(p);
+    EVP_PKEY_CTX_free(pctx);
+    OSSL_PARAM_BLD_free(tmpl);
+    OSSL_PARAM_BLD_free_params(params);
+
+    return dhpkey;
+}
+
+#  ifndef OPENSSL_NO_DEPRECATED
+/* Callback used by test_set_tmp_dh() */
+static DH *tmp_dh_callback(SSL *s, int is_export, int keylen)
+{
+    EVP_PKEY *dhpkey = get_tmp_dh_params();
+    DH *ret = NULL;
+
+    if (!TEST_ptr(dhpkey))
+        return NULL;
+
+    ret = EVP_PKEY_get1_DH(dhpkey);
+
+    EVP_PKEY_free(dhpkey);
+
+    return ret;
+}
+#  endif
+
+/*
+ * Test the various methods for setting temporary DH parameters
+ *
+ * Test  0: Default (no auto) setting
+ * Test  1: Explicit SSL_CTX auto off
+ * Test  2: Explicit SSL auto off
+ * Test  3: Explicit SSL_CTX auto on
+ * Test  4: Explicit SSL auto on
+ * Test  5: Explicit SSL_CTX auto off, custom DH params via EVP_PKEY
+ * Test  6: Explicit SSL auto off, custom DH params via EVP_PKEY
+ *
+ * The following are testing deprecated APIs, so we only run them if available
+ * Test  7: Explicit SSL_CTX auto off, custom DH params via DH
+ * Test  8: Explicit SSL auto off, custom DH params via DH
+ * Test  9: Explicit SSL_CTX auto off, custom DH params via callback
+ * Test 10: Explicit SSL auto off, custom DH params via callback
```

Comment:
I would like to see some tests that confirm that we select the correct DH parameters depending on the certificate/cipher.

Reply:
I've added a new test in sslapitest.c for this.